### DISCUSSION TITLE
Advent of Code 2024 with duck-powered SQL

### DISCUSSION THREAD

**Discussion by LennartH** [duckdb/duckdb#15490] [Show and tell]
This year, I decided to do the [Advent of Code](https://adventofcode.com/about) — an advent calendar of small programming puzzles — using SQL. I chose SQL partly as a challenge, partly to brush up on (admittedly rusty) SQL skills, but mostly out of curiosity about how well these types of problems can be solved with SQL. As it turns out, quite well, as I managed to solve all 25 days this year (a first for me)!

I've outlined the _experience_ in [my repo](https://github.com/LennartH/advent-of-code/tree/main/2024#2024-sql-with-duckdb), but I wanted to share my DuckDB-specific thoughts here. I'll try to keep it brief (let's see how that goes), so if anything is unclear or you'd like to know more, feel free to ask.

I picked DuckDB as my "runtime", because it had been mentioned a few times at work as something cool to explore and its lightweight setup (no external dependencies, in-memory mode) looked like an excellent fit for AoC. And it was. DuckDBs out of the box performance is impressive, especially considering my lack of expertise and the convoluted [monstrosities I had to conjure](https://github.com/LennartH/advent-of-code/blob/a3cf1adf00a1d9e10c821e507998d9d800bf82f6/2024/day-12_garden-groups/solution.sql#L163-L273) for some days. All without a single explicitely created index, though I'm curious whether manual indexes could improve some of the longer-running solutions ([all runtimes](https://github.com/LennartH/advent-of-code/tree/main/2024#runtimes)).

The capabilities offered by _friendly SQL_ are a game-changer — it's like night and day. They will be sorely missed when working with other dialects (shoutout to `FROM`-first syntax and trailing commas). A small disclaimer: I deliberately avoided heavy use of `MACRO`s, lambda expressions and list comprehensions. While powerful, leaning on them just felt a bit like cheating, although it might be fun to go nuts and rewrite some solutions as a follow-up.

I was expecting more "severe" difficulties or to hit a wall somewhere around day 20. It certainly wasn't easy and at times downright _brain-melting_ — [day 21](https://github.com/LennartH/advent-of-code/commits/f448e1166d9805e763a45414f0561e26788472c0/2024/day-21_keypad-conundrum/solution.sql?since=2024-12-21&until=2024-12-26) nearly broke me — but there always was another angle to explore or a different approach to try. On the flip side, some solutions felt remarkably smooth and satisfying (e.g. [day 10](https://github.com/LennartH/advent-of-code/blob/a3cf1adf00a1d9e10c821e507998d9d800bf82f6/2024/day-10_hoof-it/solution.sql) and [day 22](https://github.com/LennartH/advent-of-code/blob/a3cf1adf00a1d9e10c821e507998d9d800bf82f6/2024/day-22_monkey-market/solution.sql)). My personal highlight was day 15, a 2D puzzle about a robot pushing around boxes in a warehouse. It's responsible for nearly 60% of my total runtime, taking ~4 minutes out of the ~7 minutes overall. My solution features a [100 lines long nested recursive CTE](https://github.com/LennartH/advent-of-code/blob/d55b86a783a5a1ccb3fd8d9d34dc7a591f2c6235/2024/day-15_warehouse-woes/solution.sql#L197-L301) to handle the boxes being two tiles wide for part 2 (so `[]` instead of `o`). For me, this puzzle struck the perfect balance between incremental progress and insanity.

All in all, I had an absolute blast. I've never been this excited to work with SQL as I have over the past few weeks.

---

Now onto the promised DuckDB-specifc thoughts. I've gathered some observations and questions that I'd love to discuss (please let me know if I should create separate topics for any of these).

**Local Setup & Workflow**

- **[File Access]** I use VSCodium as my IDE and run scripts with `duckdb < solution.sql`. Each day has its own folder containing `solution.sql` and my puzzle input file. As far as I can tell file access is relative to the working directory, so I have to `cd` into the appropriate folder before running the solution. To simplify this, I built a [small bash script](https://github.com/LennartH/advent-of-code/blob/main/2024/run.sh) to execute the queries from anywhere. Is there a way to make file paths relative to the SQL file instead?
- **[Interactive Workflow]** When working on a puzzle I had the SQL file open in an editor alongside a terminal running an interactive DuckDB session, using a shortcut to paste the text selected in the editor directly into the terminal. This enabled a quick, iterative workflow, allowing simple explorative queries in the terminal while working on complex queries in the IDE. Two questions:
  - Is it possible to customize the shortcuts in the interactive session? For example, adding shortcuts to move to the start or end of a line.
  - Are there any VS Code extensions for syntax highlighting, autocompletion, or error checking? I found [DuckDB SQL Tools](https://marketplace.visualstudio.com/items?itemName=RandomFractalsInc.duckdb-sql-tools), but had issues setting up a connection and didn't notice any changes in the editor without.
- **[Persisted Variables]** In [my template](https://github.com/LennartH/advent-of-code/blob/main/template/duckdb/day-%7B%7Bday%7D%7D_%7B%7Btitle%7D%7D/solution.sql) I use a variable to toggle between example and real input when calculating results. I noticed that variables don't seem to be persisted, so views relying on variables (e.g., `FROM query_table(getvariable('mode'))`) fail with a `Binder Error: Cannot use NULL as function argument` when opening a persistent database in a new session. Is this intentional?

**Debugging & Troubleshooting**

One challenge was the lack of tools like debuggers to step through queries or inspect intermediate states, which made troubleshooting complex queries (especially recursive CTEs) tricky. For example:
```sql
WITH RECURSIVE
    shortest_paths AS (
        -- initialization
        UNION ALL (
            WITH
                -- 50 lines of helper CTEs
                im_not_working AS (
                    -- this is what I want to troubleshoot
                )
                -- more helper CTEs
            -- finalizing the step
        )
    )
FROM shortest_paths;
```
To introspect, I "simulate" a few steps of the recursion by rewriting it as a standard CTE, separating the initialization from the iterations and copying the latter (repeatedly):
```sql
WITH
    shortest_paths AS (
      -- initial statement
    ),
    shortest_pathss AS (
      -- first step, queries shortest_paths
    ),
    shortest_pathsss AS (
      -- second step, queries shortest_pathss
    ),
    -- ...
    shortest_pathssssss AS (
        WITH
            -- 50 lines of helper CTEs
            im_not_working AS (
                -- this is what I want to troubleshoot
            )
            -- more helper CTEs
        
        FROM im_not_working
        -- original statement commented out
    )

FROM shortest_pathssssss;
```
This works but quickly becomes unwieldy and you can only output one intermediate result at a time.
- Is there a better way to debug recursive CTEs or queries with many intermediate steps?
- More generally, is there tooling to split a complex CTE into multiple views or tables to make all intermediate steps accessible?

**Friendly SQL**

- **[Trailing Commas]** Trailing commas are great, especially in the beginning as they make iterating on queries much easier. Currently, they're not supported in CTEs, cross joins or function calls and I often had to rerun queries because I missed removing or adding some comma after commenting/uncommenting parts of an expression. Extending support for trailing commas would be very helpful.
- **[`CREATE OR REPLACE`]** Types seem to be one of the few elements where `CREATE OR REPLACE` can't be used and was the only case where I hade to fall back to `DROP IF EXISTS` to be able to start from scratch by running `.read solution.sql` again.
- **[Reusable Column Aliases]**  Column aliases being reusable immediately is very convenient, but can lead to "ambiguity" if the column already exists and is replaced. For example, filtering on an alias that reuses the same name might not behave as expected:
  ```sql
  -- this doesn't filter the row
  SELECT 
    list_append(path, 'b') AS path
  FROM (SELECT ['a'] AS path)
  WHERE 'b' NOT IN path;
  -- these do
  SELECT 
    list_append(old_path, 'b') AS path
  FROM (SELECT ['a'] AS old_path)
  WHERE 'b' NOT IN path;

  FROM (
    SELECT 
      list_append(path, 'b') AS path
    FROM (SELECT ['a'] AS path)
  )
  WHERE 'b' NOT IN path;
  ```
  Is there a way to explicitly use the “new” value of an alias in such cases? If not, this might be an interesting power feature. Maybe with a special syntax on use, e.g. `WHERE 'b' NOT IN |path|` (no idea what fits nicely into SQL) or an explicit declaration like `list_append(path, 'b') AS OVERRIDES path`.

---

Now that "keeping it brief" didn't really work out, but I hope it's still digestible. Thanks for reading this wall of text and a huge thanks to everyone involved in building this amazing tool!

**Comment by Tishj** [comment]
Thanks for the write-up! That's an impressive feat and the feedback is greatly appreciated 👍 

About `[File Access]`, I think that's a good idea, we could do something similar as bash I think, where an environment variable can be used to expand to the path of the script, turning the path into an absolute path:
`$SCRIPT/path/to/file`

`[Interactive Workflow]` I don't really have an answer, I don't think shortcuts can be configured as far as I'm aware.

`[Persisted Variables]` I'm not sure that's intended, but I do expect this behavior, it's stored in a session local client config and other session settings are not persisted either.

`[Trailing Commas]` I can't be entirely certain on this, but it sounds like some other issues we've covered in the past and they have all boiled down to a single problem: the parser. With the current yacc based parser it's not possible to add support for trailing commas to many places due to shift-reduce conflicts, there are plans to make to a [PEG parser](https://duckdb.org/2024/11/22/runtime-extensible-parsers.html) that *should* relieve this restriction

`[CREATE OR REPLACE]` CREATE TYPE should support the `OR REPLACE` modifier, that's indeed an unnecessary limitation currently

`[Reusable Column Aliases]` I can't really speak to this, but this just sounds like it will always be ambiguous due to the declarative nature of SQL, it's not clear whether the WHERE clause should apply directly on the source or on the final projection, I think?

**Reply by LennartH** [reply]
That's nice to hear :slightly_smiling_face: 

`[File Access]` Using an environment variable makes sense, I don't think this is concern when using other clients than the command line. Allowing variable expansion in general might be a bit much, maybe a new utility function that expands all environment variables in a given string? For example `expandenv('$HOME/path/to/file')`.

`[Interactive Workflow]` Digging in the repository I saw that [actions in the interactive shell](https://github.com/duckdb/duckdb/blob/main/tools/shell/linenoise/linenoise.cpp#L1147) are hardcoded. Not sure, but it would probably be some work to generalize this to allow configuration of the action bindings with some kind of config file (e.g. [like emacs](https://www.gnu.org/software/emacs/manual/html_node/emacs/Init-Rebinding.html#Init-Rebinding)). The CLI probably isn't the primary client, so I can understand why this wouldn't be a priority.

`[Trailing Commas]` That sounds tricky. I would have thought that parsers also got some attention in the past. This explains why syntax error messages are rather raw. But it's great to hear, that you're working on changing that.

`[Reusable Column Aliases]` Exactly, allowing aliases to be used in the scope they're declared in inevitably introduces this ambiguity (if reusing an existing alias is allowed), so anyone looking at a query needs to know which names are declared in the source to understand the details. Defaulting to the source value is reasonable, but not always what you want and the same would be the case when defaulting to the projection value. So it would be nice to be able to explicitly declare which value should be used. I think the cause of this "issue" is that you don't have an identifier for resulting projection, so adding something like `this`, `self` or `projection` could also be a solution. My example would then look like this
```sql
SELECT 
  list_append(path, 'b') AS path
FROM (SELECT ['a'] AS path)
WHERE 'b' NOT IN this.path;
```
I think that looks better than some special syntax, but no idea what caveats this could have.

**Reply by Tishj** [reply]
@LennartH <https://github.com/duckdb/duckdb/pull/9651> `getenv` already exists in the CLI 👍 

I can't really speak to the other points, maybe someone else will come along to answer some of the other points

**Reply by billwallis** [reply]
Just on reusable column aliases -- one of the few features I found useful in [SAS' SQL procedure](https://documentation.sas.com/doc/en/pgmsascdc/9.4_3.5/sqlproc/n1oihmdy7om5rmn1aorxui3kxizl.htm) was [the "calculated" keyword](https://documentation.sas.com/doc/en/pgmsascdc/9.4_3.5/sqlproc/n0jxri3yohqdwbn1slekmsc882gz.htm) which indicated that the corresponding column was a derived column. I can't remember the exact SAS semantics, but it would work something like:

```sql
from (values ('something')) as t(val)
select
    'something else' as val,

    val,             -- ambiguous, equivalent to `t.val`
    t.val,           -- explicitly the underlying column
    calculated val,  -- explicitly the calculated column
```

I'm not that opinionated on what an actual solution would look like; this SAS one is just the only one I've seen before

**Reply by LennartH** [reply]
Ah, cool to see how other systems have solved this :ok_hand: Such a keyword would a good fit as as well.

**Reply by mihirrd** [reply]
As for reusable column aliases, the ambiguity you mention makes sense, though it would be interesting to explore how other systems handle such cases. Looking forward to seeing how these evolve!

**Comment by mihirrd** [comment]
This is such an inspiring read! Tackling Advent of Code with SQL is a bold and fascinating choice, and your reflections on DuckDB’s performance and friendly SQL features are intriguing. 
What’s one DuckDB-specific feature you found most game-changing during this challenge?

**Reply by LennartH** [reply]
Thank you very much for the kind words :slightly_smiling_face: Picking a single one is hard. But if I had to I would say that DuckDB is so lightweight while still being pleasant to work with (CLI with syntax highlighting and autocompletion, nice outputs, etc.). I'm not sure if I would have seen it through with Postgres or SQLite.
